### PR TITLE
Update nikolaposa/version to version 4

### PR DESCRIPTION
**Pull request by kevin-lot**
@dontub 

I have a question.
Why do you keep the old version of nikolaposa/version (v2 and v3) ?
dependencies of nikolaposa/version are:
 - v4 => beberlei/assert and php ^7.2 (so you can upgrade your package to php 7.2, too)
 - v3 => php ^7.2 (so you can upgrade your package to php 7.2, too)
 - v2 => php ^5.5 (this time, it's shivas/versionning that locks to php 7.1)

So, to put it in a nutshell,
If you want to keep shivas/versionning compatible with php 7.1, version 2 of nikolaposa/version is mandatory.
Without it, only version 4 of nikolaposa/version is required.

### CODE REVIEW

**Comment by dontub** [comment]
I see no need to keep compatibility with PHP 7.1. When updating to nikolaposa/version v4, keeping compatibility with v2/v3 would make the code unnecessarily more complex.

**Comment by kevin-lot** [comment]
So I removed v2 and v3 of nikolaposa/version.
I tried to update travis too.

File: Formatter/GitDescribeFormatter.php [hunk 1]
```diff
@@ -15,44 +16,29 @@ class GitDescribeFormatter implements FormatterInterface
...
+        if (preg_match('/^(.*)-(\d+)-g([a-fA-F0-9]{7,40})(-dirty)?$/', $preRelease->toString(), $matches)) {
             if ((int) $matches[2] != 0) {
-                // we are not on TAG commit, add "dev" and git commit hash as pre release part
-                if (empty($matches[1])) {
-                    $version = $version->withPreRelease('dev.' . $matches[3]);
-                } else {
-                    $version = $version->withPreRelease(trim($matches[1], '-') . '.dev.' . $matches[3]);
-                }
+                // if we are not on TAG commit, add "dev" and git commit hash as pre release part
+                $withPreRelease = empty($matches[1]) ? sprintf('dev.%s',  $matches[3]) : sprintf('%s.dev.%s', trim($matches[1], '-'),  $matches[3]);
```
[the remark this answers is not in the document]
Comment (dontub) [reply Formatter/GitDescribeFormatter.php:34]: Instead of `empty` we could use `'' === $matches[1]`.

File: Formatter/GitDescribeFormatter.php [hunk 2]
```diff
@@ -15,44 +16,29 @@ class GitDescribeFormatter implements FormatterInterface
...
+                $version = $version->withPreRelease($withPreRelease);
             } else {
-                if (empty($matches[1])) {
-                    $version = $this->clearPreRelease($version);
-                } else {
-                    $version = $version->withPreRelease(trim($matches[1], '-'));
-                }
+                $withPreRelease = empty($matches[1]) ? null : trim($matches[1], '-');
```
Comment (dontub) [inline Formatter/GitDescribeFormatter.php:37]: I'd suggest:
```
elseif ('' === $matches[1]) {
    $version = $version->withPreRelease(null);
} else {
    $version = $version->withPreRelease(trim($matches[1], '-'));
}
```

File: .gitignore [hunk 3]
```diff
@@ -4,3 +4,4 @@
 /bin/
 /composer.lock
 /vendor/
+/docker-compose.yml
```
Comment (dontub) [inline .gitignore:7]: > Patterns which are specific to a particular repository but which do not need to be shared with other related repositories (e.g., auxiliary files that live inside the repository but are specific to one user’s workflow) should go into the `$GIT_DIR/info/exclude` file.

https://git-scm.com/docs/gitignore

File: .travis.yml [hunk 4]
```diff
@@ -18,13 +18,14 @@ matrix:
...
     include:
           # Minimum supported dependencies with the latest and oldest PHP version
-        - php: 7.4
+        - php: 7.2
           env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest"
-        - php: 7.1
+        - php: 7.3
+          env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest"
+        - php: 7.4
           env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest"
```
Comment (dontub) [inline .travis.yml:26]: Replace the lines above with:
```
- php: 7.4
  env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest    
- php: 7.2
  env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest"
```

**Review by dontub — CHANGES_REQUESTED** [review]
`VersionBumpCommand` needs to be changed, too.

**Comment by kevin-lot** [comment]
Sorry about the "else" "elseif".
This is job conditionning, in my company, we don't allow to code with "else" or "elseif".

**Comment by dontub** [comment]
I'd suggest to add a method `private function transformPreRelease(string $preRelease): ?string { ... }`. This would allow to immediately `return` and thus to avoid the `else`s. WDYT?

Do you have time to update `VersionBumpCommand`, too?

**Comment by kevin-lot** [comment]
> Do you have time to update `VersionBumpCommand`, too?

Done.

**Comment by kevin-lot** [comment]
> I'd suggest to add a method `private function transformPreRelease(string $preRelease): ?string { ... }`. This would allow to immediately `return` and thus to avoid the `else`s. WDYT?

I have to take a look to do with methods.
But this is a good pattern to call a method to return the good result.

File: Command/VersionBumpCommand.php [hunk 5]
```diff
@@ -62,44 +62,42 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
             $formatter = $this->manager->getFormatter();
             if ($formatter instanceof FormatterInterface) {
-                $output->writeln(sprintf('Formatter: <comment>%s</comment>', get_class($formatter)));
+                $textOutput = get_class($formatter);
```
Comment (dontub) [inline Command/VersionBumpCommand.php:65]: Actually I cannot see a benefit of assigning the formatter type to a variable here. So revert this change or convince me that it makes sense :-)

**Review by dontub — CHANGES_REQUESTED** [review]

File: Command/VersionBumpCommand.php (hunk 5 above)
Comment (kevin-lot) [reply Command/VersionBumpCommand.php]: I reverted it.

I have done that to avoid a repetition: `'Formatter: <comment>%s</comment>'`.

**Review by dontub — APPROVED** [review]

**Comment by dontub** [comment]
Thanks for contribution @kevin-lot!
